File: src/stages/dialog.ts
import { createInterface } from 'node:readline';
import chalk from 'chalk';
import type { Ambiguity } from '../ir/types.js';
import type { SoelConfig } from '../config.js';
import { llmRequest, extractJSON } from '../llm/openrouter.js';
import { loadPrompt } from '../llm/prompts.js';
import { log } from '../utils/logger.js';

/**
 * Interactive dialog loop to resolve ambiguities with the user.
 * Returns resolved ambiguities with user's choices.
 */
export async function resolveAmbiguities(
  ambiguities: Ambiguity[],
  sourceText: string,
  config: SoelConfig
): Promise<Ambiguity[]> {
  if (ambiguities.length === 0) {
    log.info('No ambiguities to resolve');
    return ambiguities;
  }

  log.stage('Dialogical ambiguity resolution');
  console.error(
    chalk.dim(`\n  ${ambiguities.length} ambiguities need resolution. `) +
    chalk.dim(`(max ${config.compiler.maxDialogRounds} rounds)\n`)
  );

  const rl = createInterface({
    input: process.stdin,
    output: process.stderr,
  });

  const ask = (prompt: string): Promise<string> =>
    new Promise((resolve) => rl.question(prompt, resolve));

  const resolved: Ambiguity[] = [];
  let round = 0;

  for (const amb of ambiguities) {
    if (round >= config.compiler.maxDialogRounds) {
      log.warn(`Max dialog rounds (${config.compiler.maxDialogRounds}) reached, auto-resolving remaining`);
      // Auto-resolve remaining with highest confidence option
      resolved.push(autoResolve(amb));
      continue;
    }

    round++;
    console.error(chalk.yellow(`\n─── Ambiguity ${amb.id} [${amb.category}] ───`));
    console.error(chalk.white(`  ${amb.description}\n`));

    // Show options
    for (let i = 0; i < amb.options.length; i++) {
      const opt = amb.options[i];
      const conf = chalk.dim(`(${(opt.confidence * 100).toFixed(0)}%)`);
      console.error(`  ${chalk.cyan(`${i + 1})`)} ${opt.label} ${conf}`);
      if (opt.description !== opt.label) {
        console.error(chalk.dim(`     ${opt.description}`));
      }
    }
    console.error(`  ${chalk.cyan(`${amb.options.length + 1})`)} ${chalk.italic('Custom response...')}`);

    const answer = await ask(chalk.green('\n  Your choice: '));
    const choice = parseInt(answer.trim(), 10);

    if (choice >= 1 && choice <= amb.options.length) {
      const chosen = amb.options[choice - 1];
      resolved.push({
        ...amb,
        resolution: {
          chosen: chosen.label,
          rationale: `User selected: ${chosen.description}`,
        },
      });
      console.error(chalk.green(`  ✓ Resolved: ${chosen.label}`));
    } else if (choice === amb.options.length + 1 || isNaN(choice)) {
      // Custom response — use LLM to interpret
      const customInput = isNaN(choice)
        ? answer.trim()
        : await ask(chalk.green('  Describe your preference: '));

      if (customInput.trim()) {
        const interpretation = await interpretCustomResponse(
          amb,
          customInput,
          sourceText,
          config
        );
        resolved.push({
          ...amb,
          resolution: {
            chosen: interpretation,
            rationale: `User provided custom input: "${customInput}"`,
          },
        });
        console.error(chalk.green(`  ✓ Resolved: ${interpretation}`));
      } else {
        resolved.push(autoResolve(amb));
        console.error(chalk.dim(`  → Auto-resolved with highest confidence option`));
      }
    } else {
      resolved.push(autoResolve(amb));
      console.error(chalk.dim(`  → Auto-resolved with highest confidence option`));
    }
  }

  rl.close();
  log.success(`Resolved ${resolved.filter((a) => a.resolution).length}/${ambiguities.length} ambiguities`);
  return resolved;
}

function autoResolve(amb: Ambiguity): Ambiguity {
  const best = amb.options.reduce((a, b) =>
    a.confidence >= b.confidence ? a : b
  );
  return {
    ...amb,
    resolution: {
      chosen: best.label,
      rationale: `Auto-resolved: highest confidence option (${(best.confidence * 100).toFixed(0)}%)`,
    },
  };
}

async function interpretCustomResponse(
  amb: Ambiguity,
  userInput: string,
  sourceText: string,
  config: SoelConfig
): Promise<string> {
  try {
    const systemPrompt = loadPrompt('ambiguity-resolver');
    const raw = await llmRequest({
      apiKey: config.openrouter.apiKey!,
      model: config.openrouter.model,
      messages: [
        { role: 'system', content: systemPrompt },
        {
          role: 'user',
          content: JSON.stringify({
            ambiguity: amb,
            user_response: userInput,
            source_context: sourceText.slice(0, 2000),
          }),
        },
      ],
      temperature: 0.2,
      maxTokens: 512,
    });

    const json = extractJSON(raw);
    const parsed = JSON.parse(json) as { interpretation: string };
    return parsed.interpretation ?? userInput;
  } catch {
    return userInput;
  }
}

/**
 * Auto-resolve all ambiguities without user interaction.
 */
export function autoResolveAll(ambiguities: Ambiguity[]): Ambiguity[] {
  return ambiguities.map(autoResolve);
}

File: src/Soel/Stages/Dialog.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Soel.Stages.Dialog
  ( resolveAmbiguities
  , autoResolveAll
  , autoResolve
  ) where

import Control.Monad.IO.Class (liftIO)
import Control.Monad.Reader (ask)
import Data.Aeson (eitherDecode', encode, Value(..), (.:), object, (.=), toJSON)
import Data.Aeson.Types (parseMaybe)
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Char8 as BLC
import qualified Data.Text.Encoding as TE
import Data.List (maximumBy)
import Data.Maybe (isJust)
import Data.Ord (comparing)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.Console.Haskeline (runInputT, defaultSettings, getInputLine)
import System.IO (stderr, hFlush)
import Text.Read (readMaybe)

import Soel.Types (App, SoelConfig(..), requireApiKeyM)
import Soel.Config (CompilerConfig(..), OpenRouterConfig(..))
import Soel.IR.Types
import Soel.IR.Validate ()  -- import ToJSON/FromJSON orphan instances
import Soel.LLM.OpenRouter (llmRequest, extractJSON, LLMRequestOptions(..), LLMMessage(..), LLMRole(..))
import Soel.LLM.Prompts (getPrompt)
import Soel.Utils.Logger (logStage, logInfo, logWarn, logSuccess)

-- | Interactive dialog loop to resolve ambiguities with the user.
-- Returns resolved ambiguities with user's choices.
resolveAmbiguities :: [Ambiguity] -> Text -> App [Ambiguity]
resolveAmbiguities ambs sourceText = do
  if null ambs
    then do
      liftIO $ logInfo "No ambiguities to resolve"
      pure ambs
    else do
      cfg <- ask
      let maxRounds = ccMaxDialogRounds (cfgCompiler cfg)

      liftIO $ logStage "Dialogical ambiguity resolution"
      liftIO $ TIO.hPutStr stderr $
        "\n  " <> T.pack (show (length ambs)) <> " ambiguities need resolution. "
        <> "(max " <> T.pack (show maxRounds) <> " rounds)\n\n"
      liftIO $ hFlush stderr

      go cfg 0 maxRounds ambs []
  where
    go :: SoelConfig -> Int -> Int -> [Ambiguity] -> [Ambiguity] -> App [Ambiguity]
    go _   _     _         []         resolved = do
      let resCount = length (filter (isJust . ambResolution) resolved)
          total    = length resolved
      liftIO $ logSuccess $ "Resolved " <> T.pack (show resCount)
                         <> "/" <> T.pack (show total) <> " ambiguities"
      pure (reverse resolved)
    go cfg rnd maxRounds (amb:rest) resolved
      | rnd >= maxRounds = do
          liftIO $ logWarn $ "Max dialog rounds (" <> T.pack (show maxRounds)
                          <> ") reached, auto-resolving remaining"
          let remaining = map autoResolve (amb:rest)
          let resCount = length (filter (isJust . ambResolution) (resolved ++ remaining))
              total    = length resolved + length remaining
          liftIO $ logSuccess $ "Resolved " <> T.pack (show resCount)
                             <> "/" <> T.pack (show total) <> " ambiguities"
          pure (reverse resolved ++ remaining)
      | otherwise = do
          -- Show the ambiguity
          liftIO $ TIO.hPutStr stderr $
            "\n--- Ambiguity " <> ambId amb <> " [" <> categoryText (ambCategory amb) <> "] ---\n"
            <> "  " <> ambDescription amb <> "\n\n"
          liftIO $ hFlush stderr

          -- Show numbered options
          liftIO $ mapM_ (\(i, opt) -> do
            let conf = T.pack (show (round (aoConfidence opt * 100) :: Int)) <> "%"
            TIO.hPutStr stderr $
              "  " <> T.pack (show i) <> ") " <> aoLabel opt <> " (" <> conf <> ")\n"
            if aoDescription opt /= aoLabel opt
              then TIO.hPutStr stderr $ "     " <> aoDescription opt <> "\n"
              else pure ()
            ) (zip ([1..] :: [Int]) (ambOptions amb))

          let customIdx = length (ambOptions amb) + 1
          liftIO $ TIO.hPutStr stderr $
            "  " <> T.pack (show customIdx) <> ") Custom response...\n"
          liftIO $ hFlush stderr

          -- Get user input via haskeline
          mAnswer <- liftIO $ runInputT defaultSettings $ getInputLine "\n  Your choice: "

          case mAnswer of
            Nothing -> do
              -- EOF / no input => auto-resolve
              let resolved' = autoResolve amb
              liftIO $ TIO.hPutStr stderr "  -> Auto-resolved with highest confidence option\n"
              liftIO $ hFlush stderr
              go cfg (rnd + 1) maxRounds rest (resolved' : resolved)

            Just answer -> do
              let trimmed = T.strip (T.pack answer)
                  mChoice = readMaybe answer :: Maybe Int

              case mChoice of
                Just choice
                  | choice >= 1 && choice <= length (ambOptions amb) -> do
                      -- User selected a numbered option
                      let chosen = ambOptions amb !! (choice - 1)
                          resolved' = amb
                            { ambResolution = Just AmbiguityResolution
                                { arChosen    = aoLabel chosen
                                , arRationale = "User selected: " <> aoDescription chosen
                                }
                            }
                      liftIO $ TIO.hPutStr stderr $ "  -> Resolved: " <> aoLabel chosen <> "\n"
                      liftIO $ hFlush stderr
                      go cfg (rnd + 1) maxRounds rest (resolved' : resolved)

                  | choice == customIdx -> do
                      -- User wants custom response, prompt for it
                      mCustom <- liftIO $ runInputT defaultSettings $
                        getInputLine "  Describe your preference: "
                      handleCustomInput cfg amb rnd maxRounds rest resolved mCustom

                _ -> do
                  -- Not a valid number — treat as custom text input
                  if T.null trimmed
                    then do
                      let resolved' = autoResolve amb
                      liftIO $ TIO.hPutStr stderr "  -> Auto-resolved with highest confidence option\n"
                      liftIO $ hFlush stderr
                      go cfg (rnd + 1) maxRounds rest (resolved' : resolved)
                    else
                      handleCustomInput cfg amb rnd maxRounds rest resolved (Just answer)

    handleCustomInput :: SoelConfig -> Ambiguity -> Int -> Int
                      -> [Ambiguity] -> [Ambiguity] -> Maybe String -> App [Ambiguity]
    handleCustomInput cfg amb rnd maxRounds rest resolved mInput = case mInput of
      Nothing -> do
        let resolved' = autoResolve amb
        liftIO $ TIO.hPutStr stderr "  -> Auto-resolved with highest confidence option\n"
        liftIO $ hFlush stderr
        go cfg (rnd + 1) maxRounds rest (resolved' : resolved)
      Just input -> do
        let customText = T.strip (T.pack input)
        if T.null customText
          then do
            let resolved' = autoResolve amb
            liftIO $ TIO.hPutStr stderr "  -> Auto-resolved with highest confidence option\n"
            liftIO $ hFlush stderr
            go cfg (rnd + 1) maxRounds rest (resolved' : resolved)
          else do
            interpretation <- interpretCustomResponse amb customText sourceText cfg
            let resolved' = amb
                  { ambResolution = Just AmbiguityResolution
                      { arChosen    = interpretation
                      , arRationale = "User provided custom input: \"" <> customText <> "\""
                      }
                  }
            liftIO $ TIO.hPutStr stderr $ "  -> Resolved: " <> interpretation <> "\n"
            liftIO $ hFlush stderr
            go cfg (rnd + 1) maxRounds rest (resolved' : resolved)

-- | Use the LLM to interpret a custom user response for an ambiguity.
interpretCustomResponse :: Ambiguity -> Text -> Text -> SoelConfig -> App Text
interpretCustomResponse amb userInput sourceText cfg = do
  let systemPrompt = getPrompt "ambiguity-resolver"

  apiKey <- requireApiKeyM

  let userContentVal = ambiguityContext amb userInput sourceText
      userContent = T.pack $ BLC.unpack $ encode userContentVal

  result <- llmRequest LLMRequestOptions
    { lroApiKey      = apiKey
    , lroModel       = orModel (cfgOpenRouter cfg)
    , lroMessages    =
        [ LLMMessage RoleSystem systemPrompt
        , LLMMessage RoleUser userContent
        ]
    , lroTemperature = 0.2
    , lroMaxTokens   = Nothing
    , lroJsonMode    = False
    }

  let jsonText = extractJSON result
  case eitherDecode' (BL.fromStrict $ TE.encodeUtf8 jsonText) of
    Right (Object obj) -> case parseMaybe (.: "interpretation") obj of
      Just (String interp) -> pure interp
      _                    -> pure userInput
    _ -> pure userInput

  where
    ambiguityContext :: Ambiguity -> Text -> Text -> Value
    ambiguityContext a ui src = object
      [ "ambiguity"      .= toJSON a
      , "user_response"  .= ui
      , "source_context" .= T.take 2000 src
      ]

-- | Auto-resolve a single ambiguity by picking the highest-confidence option.
autoResolve :: Ambiguity -> Ambiguity
autoResolve amb
  | null (ambOptions amb) = amb
  | otherwise =
      let best = maximumBy (comparing aoConfidence) (ambOptions amb)
          pct  = T.pack (show (round (aoConfidence best * 100) :: Int)) <> "%"
      in  amb
            { ambResolution = Just AmbiguityResolution
                { arChosen    = aoLabel best
                , arRationale = "Auto-resolved: highest confidence option (" <> pct <> ")"
                }
            }

-- | Auto-resolve all ambiguities without user interaction.
autoResolveAll :: [Ambiguity] -> [Ambiguity]
autoResolveAll = map autoResolve

